<a href="https://colab.research.google.com/github/Solo2705/Ybi_projects/blob/main/WineQualityPrediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Wine Quality Testing

It has many columns :
* fixed acidity
*volatile acidity
*citric acid
*residual sugar
*chlorides
*free sulfur dioxide
*total sulfur dioxide
*density
*pH
*sulphates
*alcohol
*quality



## OBJECTIVE
Objective of this model is to predict quality of wine using the above parameters.

## IMPORTING LIBRARIES

In [ ]:
import pandas as pd
import numpy as np


## IMPORT DATA

In [ ]:
wine_df = pd.read_csv(r'https://raw.githubusercontent.com/YBIFoundation/Dataset/main/WhiteWineQuality.csv',sep=';')


## DESCRIBE DATA

In [ ]:
wine_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [ ]:
wine_df.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    4898.000000       4898.000000  4898.000000     4898.000000   
mean        6.854788          0.278241     0.334192        6.391415   
std         0.843868          0.100795     0.121020        5.072058   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.300000          0.210000     0.270000        1.700000   
50%         6.800000          0.260000     0.320000        5.200000   
75%         7.300000          0.320000     0.390000        9.900000   
max        14.200000          1.100000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  4898.000000          4898.000000           4898.000000  4898.000000   
mean      0.045772            35.308085            138.360657     0.994027   
std       0.021848            17.007137             42.498065     0.002991   
min       0.009000             2.000000              9.000000     0.987110   
25%       0.036000            23.000000            108.000000     0.991723   
50%       0.043000            34.000000            134.000000     0.993740   
75%       0.050000            46.000000            167.000000     0.996100   
max       0.346000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  4898.000000  4898.000000  4898.000000  4898.000000  
mean      3.188267     0.489847    10.514267     5.877909  
std       0.151001     0.114126     1.230621     0.885639  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.090000     0.410000     9.500000     5.000000  
50%       3.180000     0.470000    10.400000     6.000000  
75%       3.280000     0.550000    11.400000     6.000000  
max       3.820000     1.080000    14.200000     9.000000

In [ ]:
wine_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


In [ ]:
wine_df.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality'],
      dtype='object')

In [ ]:
wine_df.shape

(4898, 12)

In [ ]:
wine_df['quality'].value_counts()

quality
6    2198
5    1457
7     880
8     175
4     163
3      20
9       5
Name: count, dtype: int64

## Define Target Variable (y) and Feature Variables (X)

In [ ]:
y = wine_df['quality']
X = wine_df.drop(['quality'],axis=1)

## STANDARDIZE DATA

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
ss = StandardScaler()
X = ss.fit_transform(X)

## TRAIN TEST SPLIT

In [ ]:
from sklearn.model_selection import train_test_split


In [ ]:
X_train ,X_test , y_train , y_test = train_test_split(X,y,test_size=0.3,random_state=2529)


## MODELING

In [ ]:
from sklearn.svm import SVC

model = SVC()
model.fit(X_train , y_train)

SVC()

In [ ]:
y_pred = model.predict(X_test)

## MODEL EVALUATION

In [ ]:
from sklearn.metrics import confusion_matrix , classification_report,accuracy_score

In [ ]:
confusion_matrix(y_test , y_pred)

array([[  0,   0,   2,   2,   0,   0],
       [  0,   1,  38,  18,   0,   0],
       [  0,   2, 248, 171,   4,   0],
       [  0,   0, 133, 502,  44,   0],
       [  0,   0,  10, 174,  73,   0],
       [  0,   0,   1,  33,  14,   0]])

In [ ]:
accuracy_score(y_test , y_pred)

0.5605442176870749

In [ ]:
print(classification_report(y_test , y_pred))

              precision    recall  f1-score   support

           3       0.00      0.00      0.00         4
           4       0.33      0.02      0.03        57
           5       0.57      0.58      0.58       425
           6       0.56      0.74      0.64       679
           7       0.54      0.28      0.37       257
           8       0.00      0.00      0.00        48

    accuracy                           0.56      1470
   macro avg       0.33      0.27      0.27      1470
weighted avg       0.53      0.56      0.53      1470



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## RE-RUN WITH TWO GROUPS OF WINE QUALITY

In [ ]:
y = wine_df['quality'].apply(lambda y_value:1 if y_value>=6 else 0)
y.value_counts()

quality
1    3258
0    1640
Name: count, dtype: int64

## TRAIN TEST SPLIT

In [ ]:
from sklearn.model_selection import train_test_split
X_train ,X_test , y_train , y_test = train_test_split(X,y,test_size=0.3,random_state=2529)


## MODELING

In [ ]:
from sklearn.svm import SVC

model = SVC()
model.fit(X_train , y_train)

SVC()

In [ ]:
y_pred = model.predict(X_test)


## MODEL EVALUATION

In [ ]:
from sklearn.metrics import confusion_matrix , classification_report,accuracy_score


In [ ]:
confusion_matrix(y_test,y_pred)

array([[289, 197],
       [137, 847]])

In [ ]:
accuracy_score(y_test,y_pred)

0.7727891156462585

In [ ]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.68      0.59      0.63       486
           1       0.81      0.86      0.84       984

    accuracy                           0.77      1470
   macro avg       0.74      0.73      0.73      1470
weighted avg       0.77      0.77      0.77      1470



## EXPLANATION
* First we import in-built libraries and dataset from github.
* We check column names and description of dataset.
* Then standardize data to make transformation easier.
* Define X(Independent) and Y(Dependent) variable.
* Split dataset into train and test data.
* Model selection: We choose Support Vector Classifier.
* After implementation we do model evaluation

In this way accuracy was 56% so we need to do it other way



So we re-run the same code by grouping quality of wine into two


groups. Now the accuracy using Support Vector Classifiers is 77%